### ISSUE TITLE
Weird impossibly high spike in "Solar production" on Home Assistant energy dashboard

### ISSUE THREAD

**Issue by ncot-tech** [hultenvp/solis-sensor#296]
I don't know if this is an issue with this integration or with Home Assistant, but recently I've noticed my solar production graph has a weird and impossibly high spike in the morning...

![image](https://github.com/hultenvp/solis-sensor/assets/7870113/c443dacb-3cff-444c-be95-ce26079f3093)

(I mean, I'd like to somehow magically make that much energy in the morning from my 3kW setup, I could charge my car with it! but I think something's not quite right)

However, the sensor it's using looks fine...

![image](https://github.com/hultenvp/solis-sensor/assets/7870113/fe58809e-e206-45ab-adf5-38acd76c77f0)

And it matches what Solis Cloud thinks

![image](https://github.com/hultenvp/solis-sensor/assets/7870113/96cd7d2a-15fd-4c4e-97c3-88436ae31ee8)

So the data seems to be coming out of Solis Cloud properly, but for some reason HA is going a bit strange. Any ideas how I can debug this? My energy dashboard is set up like this:

![image](https://github.com/hultenvp/solis-sensor/assets/7870113/f3f2717f-46a8-4d7f-a575-062394a882ad)

**Comment by bluemoon83** [comment]
I've had similar issues over the couple of weeks but mine does it between 2300-0000 and the spike is the size of solar production for that day. 
<img width="1418" alt="Screenshot 2023-06-15 at 21 01 00" src="https://github.com/hultenvp/solis-sensor/assets/48335247/36a94f84-18fd-47aa-8a9b-d56f1f9a438d">
<img width="1107" alt="Screenshot 2023-06-15 at 21 01 47" src="https://github.com/hultenvp/solis-sensor/assets/48335247/573eb390-a76d-4baf-bc36-31945bbb219e">

Not sure why it is happening this post https://community.home-assistant.io/t/correcting-incorrect-energy-statistics/429720 does show how you can remove them, with relative ease, afterwards so it doesn't distort your stats.

**Comment by jahpster** [comment]
I have the same problem since the 6 of June.

**Comment by hultenvp** [comment]
See also #273 

@ncot-tech can you also check the history of the "Solis daily grid energy purchased"  and "Solis daily on grid energy". Wouldn't be surprised if anything is wrong with those. 
Soliscloud API is extremely unreliable. This is likely one of those examples, I don't think this is an energy dash issue. If I know which fields misbehave I might be able to apply some filtering to mitigate the issues, but no guarantees.

**Comment by wesjuhdabomb** [comment]
@hultenvp. Same issue here, however I only see 0 amounts on the "Solis daily grid energy purchased" and "Solis daily on grid energy".

What I do see on the days when it is spiking (not happening every day), "Solid Energy Today" jumps to 0 and back to the total production in the last hour before midnight. Hopefully this helps. In my case a filter for any production between 23:00 and 0:00 would eleminate the issue. 
![image](https://github.com/hultenvp/solis-sensor/assets/9979893/eb4a2835-4f87-48ce-97f2-4454e12b5606)

**Comment by bbartje** [comment]
> @hultenvp. Same issue here, however I only see 0 amounts on the "Solis daily grid energy purchased" and "Solis daily on grid energy".
> 
> What I do see on the days when it is spiking (not happening every day), "Solid Energy Today" jumps to 0 and back to the total production in the last hour before midnight. Hopefully this helps. In my case a filter for any production between 23:00 and 0:00 would eleminate the issue. ![image](https://user-images.githubusercontent.com/9979893/248903981-eb4a2835-4f87-48ce-97f2-4454e12b5606.png)

That sounds like a perfect workaround, I have the same problem and don't want to keep cleaning up my stats.
How do you create such a filter?

**Comment by chmutoff** [comment]
I have exactly the same issue. The counter in solis resets a 0:00 and this seems to break the stats in energy dashboard

**Comment by hultenvp** [comment]
Others have reported these fluctuations on different moments during the day.
Filtering out could work under the following conditions:
* The integration has no ability to persist state, so you'll never be able to filter out if HA restarts just at the moment SC reports 0 kWh.
* There is already a filter to ignore any non-zero values when the inverter is still off. This is to filter out situations where energy is reset too late after midnight. This filter and the reset at midnight should continue to work.

The existing filter is in [service.py](https://github.com/hultenvp/solis-sensor/blob/ee36537f37e80d5593f73483ec45bcac230d48c1/custom_components/solis/service.py#L158)
A new one must be added that keeps track of previous values and ignores drops (any drop, even if it's not to 0 is not allowed if it's not the daily reset)

As said in different threads I have little time at the moment to work on the integration. I do welcome and will process community contributions.

**Comment by wesjuhdabomb** [comment]
Not fully understanding the code, but I made a quick and dirty change in my homeassistant setup and it seems the spikes are gone ever since. The change was made 1wk ago and I used to see them every 2-3 days.

From my basic understanding the code, my reasoning was to ignore any values from polluting the data when the inverter is off, since there should not be any additional energy production with the inverter in OFF state. These lines were added after line 163 in service.py.

```
is_pm = datetime.now().hour > 12
if getattr(data, INVERTER_STATE) == 2 and is_pm:
   return
```
